### ISSUE TITLE
As a korifi operator I want to be able to use my own log cache endpoint

### ISSUE THREAD

**Issue by georgethebeatle** [cloudfoundry/korifi#3668]
### Dev Notes

- Currently korifi implements the `api/v1/read` log cache endpoint in order to get logs for app
- Currently korifi queries the metrics servce crds for pod stats in order to satisfy the processes/stats endpoint
- In cf for vms process stats are also [fetched from the log cache api](https://github.com/cloudfoundry/cloud_controller_ng/blob/431ea8bf31948a963d87fe5c6cd557a1ebc7506c/lib/cloud_controller/diego/reporters/instances_stats_reporter.rb#L36).
- We should implements the `api/v1/query` in the log cache handler and make the process stats repository loop back to this endpoint
- Furthermore we should also make the logcache api endpoint configurable in the helm chart so that operators can override it
- If no value is provided we will keep loooping back to the naive implementation by default
- [Log cache api spec](https://github.com/cloudfoundry/log-cache-release/blob/main/src/README.md)

**Comment by chombium** [comment]
Hi,

CF for VMs suports various [log types](https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html) which practically are the app logs themselves combined with the logs written from the platform components and a related to an app. An example would be: access logs, staging logs, app restarts, rescheduling and similar. All these logs are stored in Log Cache and are available through the Log Cache API. It would be nice to have all these app and app lifecycle logs available, but we should find out how and where to get them from.

I've tested and compared what kind of logs are being printed out with `cf logs` in CF-for-VMs and Korifi in/after different cf lifecycle events. __Note:__ `cf logs` practically calls the Log Cache `v1/read` api for the given app with `limit=200` parameter.  

During my tests I've seen so far `cf push` output in Korifi is pretty similar if not identical to the one on CF-for-VMs. For the other commands like `cf restage`, `cf restart` Korifi only outputs the app logs and the other platform related logs have to be fetched via the pod events `kubectl get events <pod>`. In Korifi we have to practically combine data from various places to get a "cf logs" output as close to what we have on CF-for-VMs.

The other aspect to think about is that CF for VMs follows the [Loggregator API](https://github.com/cloudfoundry/loggregator-api) format and observability metadata, if we want to use the same format and where and how do we get the data from.

In regards to the `api/v1/query` endpoint it is Prometheus compatible API, so if can make the metrics available via Prometheus we only need to map the api endpoint somewhere.

I guess, the `/api/v1/meta` would be out of scope if we simply read the logs from the k8s API server server. We could add something if needed afterwards.

The one important thing to think about is how to collect and merge the logs and metrics in a single output(api) in case an app has multiple instances (a workload with multiple pods).

For implementation of the API we could take a look at the [log-cache-release](https://github.com/cloudfoundry/log-cache-release) and the [log-cache-cf-cli plugin](https://github.com/cloudfoundry/log-cache-cli).

btw. I'm one of the maintainers of the Loggregator, the CF's logging and metrics stack and I want to help ;)

**Comment by georgethebeatle** [comment]
@chombium thank you for your input! All your observations are correct, however we are thinking about a very minimal "batteries-included" implementation of the log-cache api to get the most basic behaviour. This of course leaves the door open for anyone to bring their own log cache that will have to do all the things you talk about in your comment. 

As of today Korifi's API already provides a minimal `GET /api/v1/read` implementation that can only read container logs for an app in oder to support the needs of the `cf push` and `cf logs` commands.

The problematic part is the fetching of app metrics. Towards the end of the `cf push` operation the cli starts querying `/v3/processes/<procees-guid>/stats`. Korifi implements this endpoint by directly querying the metrics server custom resources. This implementation makes it impossible to bring your own log cache implementation as of today.

Our idea is to push the [metrics fetching code](https://github.com/cloudfoundry/korifi/blob/739b1d8ec0e4b27ac784807aa9de3e391a323cd5/api/repositories/metrics_repository.go#L36) behind the `/api/v1/read` interface. This, by the way is exactly what the [cloud controller does](https://github.com/cloudfoundry/cloud_controller_ng/blob/431ea8bf31948a963d87fe5c6cd557a1ebc7506c/lib/cloud_controller/diego/reporters/instances_stats_reporter.rb#L36) in classic cf.

TL;DR We do not have the capacity to implement log cache properly, but we want to enable Korifi users to bring their own.

**Comment by danail-branekov** [comment]
Hi @chombium 

> If we implement the Log Cache API fasade properly it will be easy for someone to get their own backend

The log cache REST API **IS** the facade. The way we envision it is that we would have a helm value where users would be able to specify where their real, fancy, open-telemetry-based, whatever logcache implementation is. As @georgethebeatle said, we do not have the capacity to implement logcache properly. 

Furthermore, Korifi is a CF API implementation, therefore logcache should not be implemented at all in the first place. We have this very basic and naive "batteries-included" implementation to only facilitate cases where users just want to play with Korifi (and do not want to install additional stuff), or are happy to accept abusing the metrics server.

**Comment by chombium** [comment]
Hi @danail-branekov ,

I completely agree with what you've written. We should add the basic functionality so that `cf push` and `cf logs` will work decently and we can later think and discuss about production ready things.

**Comment by georgethebeatle** [comment]
Closing in favour of the sub-issues
